eva_submission/biosamples_submission.py

```diff
@@ -229,29 +280,88 @@ def convert_sample_data_to_curation_object(self, future_sample):
 
         return dict(curation=curation_object, sample=future_sample.get('accession'))
 
-    def submit_to_bsd(self, samples_data):
+    @staticmethod
+    def _update_from_array(key, sample_source, sample_dest):
+        """Add the element of an array stored in specified key from source to destination"""
+        if key in sample_source:
+            if key not in sample_dest:
+                sample_dest[key] = []
+            for element in sample_source[key]:
+                if element not in sample_dest[key]:
+                    sample_dest[key].append(element)
+
+    def _update_samples_with(self, sample_source, sample_dest):
+        """Update a BioSample object with the value of another"""
+        for attribute in sample_source['characteristics']:
+            if attribute not in sample_dest['characteristics']:
+                sample_dest['characteristics'][attribute] = sample_source['characteristics'][attribute]
+        self._update_from_array('externalReferences', sample_source, sample_dest)
+        self._update_from_array('relationships', sample_source, sample_dest)
+        self._update_from_array('contact', sample_source, sample_dest)
+        self._update_from_array('organization', sample_source, sample_dest)
+        for key in ['taxId', 'accession', 'name', 'release']:
+            if key in sample_source and key not in sample_dest:
+                sample_dest[key] = sample_source[key]
+
+    def create_sample_to_overwrite(self, sample):
+        """Create all the sample that will be used to overwrite the exising ones"""
+        # We only add existing characteristics if we do not want to remove anything
+        if self.can_overwrite(sample) and not self.allow_removal:
+            # retrieve the sample without any curation and add the new data on top
+            current_sample = self._get_existing_sample(sample.get('accession'))
+            self._update_samples_with(current_sample, sample)
+        return sample
+
+    def create_derived_sample(self, sample):
+        skipped_attributes = ['SRA accession']
+        if self.can_derive(sample):
+            derived_sample = deepcopy(sample)
+            current_sample = self._get_existing_sample(derived_sample.get('accession'))
+            self._update_samples_with(current_sample, derived_sample)
+            # Remove the accession of previous samples
+            accession = derived_sample.pop('accession')
+            # Remove the SRA accession if it is there
+            if 'SRA accession' in derived_sample['characteristics']:
+                derived_sample['characteristics'].pop('SRA accession')
+            derived_sample['release'] = _now
+            return derived_sample, accession
+
+    def submit_biosamples_to_bsd(self, samples_data):
```

Comment:
I think I would prefer if this were split into different methods, so that it's more explicit when the method is being used which action is being taken, rather than having to refer back to how the submitter object was created. So something like `submit_samples`, `curate_samples`, and so on, with appropriate error handling if you try to execute an action that's not allowed (e.g. overwrite a sample we don't own).

Reply:
I agree that it can be refactored but there is an (unused now) feature I wanted to keep which allow a set of sample to pass through this function and be treated differently depending on the specific sample's need.
It is also the reason why the submitter can receive a tuple of operations.
One could argue that this is one of the reason why the brokering got messy and the feature is unnecessary but on the other hand I feel like it will be needed at some point.

Reply:
Got it, thanks for the explanation and I think I agree it makes sense to keep it this way for now.